### PR TITLE
core (largely doc): Differentiate message types from thread flags

### PR DESCRIPTION
**Pull request by chrysn**
### Contribution description

How to make sure own thread flags or message types do not conflict with preexisting ones (especially with the lack of a comprehensive of predefined ones for the latter) is difficult from the current documentation; this change set:

* clarifies that flags *will* be set by the OS without solicitation,
  * and adds a single define against which custom use can be checked; and
* states that messages are *not* set by the OS willy-nilly, and that system-wide uniqueness of numbers is not really necessary (but just a good practice by the standard library)

(and fixes an obsolete number in the process).

### Testing procedure

* Agree with the changes to the docs
* Watch nothing break in the tests from an added define

### Issues/PRs references

This contains clean-up from https://github.com/RIOT-OS/RIOT/pull/7717

### CODE REVIEW

File: core/include/thread_flags.h [hunk 1]
```diff
@@ -98,6 +101,20 @@ extern "C" {
...
  */
 #define THREAD_FLAG_TIMEOUT         (1u << 14)
+
+
+/**
+ * @brief Comprehensive set of all predefined flags
+ *
+ * This bit mask is set for all thread flag bits that are predefined in RIOT.
+ * Flags within this set may be set on a thread by the operating system without
+ * thre thread soliciting them (though not all are; for example, @ref
```
Comment (kaspar030) [inline core/include/thread_flags.h:111]:
```suggestion
 * the thread soliciting them (though not all are; for example, @ref
```

File: core/include/thread_flags.h [hunk 2]
```diff
@@ -98,6 +101,20 @@ extern "C" {
...
+ * THREAD_FLAG_TIMEOUT is not).
+ *
+ * When using custom flags, asserting that they are not in this set can help
+ * avoid conflict with future additions to the predefined flags.
+ */
+#define THREAD_FLAG_PREDEFINED      THREAD_FLAG_MSG_WAITING | THREAD_FLAG_TIMEOUT
```
Comment (kaspar030) [inline core/include/thread_flags.h:117]:
```suggestion
#define THREAD_FLAG_PREDEFINED      (THREAD_FLAG_MSG_WAITING | THREAD_FLAG_TIMEOUT)
```

File: core/include/thread_flags.h [hunk 3]
```diff
@@ -98,6 +101,20 @@ extern "C" {
  * @see xtimer_set_timeout_flag
  */
 #define THREAD_FLAG_TIMEOUT         (1u << 14)
+
+
```
Comment (kaspar030) [inline core/include/thread_flags.h:105]: extra newline

File: core/include/msg.h [hunk 4]
```diff
@@ -15,9 +15,20 @@
...
+ * Some message types are predefined; for example, @ref
+ * GNRC_NETAPI_MSG_TYPE_RCV & co are defined. These are fixed in the sense that
+ * registering for a particular set of messages (for the above, e.g. @ref
+ * gnrc_netreg_register) will use these message types. Threads that do nothing
+ * to receive such messages need not worry can safely use the same numbers for
+ * other purposes. The predefined types use non-conflicting ranges (e.g.
+ * `0x02NN`) to avoid ruling out simultaneous use of different components in
+ * the same thread.
+ *
+ * In general, threads may consider it an error to send them a message they did
```
Comment (kaspar030) [inline core/include/msg.h:30]: Hm, best practice is to maybe log but ignore unexpected messages. Would this sentence make users error out instead?
Maybe add the best practice here.

File: core/include/msg.h [hunk 5]
```diff
@@ -15,9 +15,20 @@
...
- * should be set to a system-wide unique value. The content can either be
- * provided as a 32-bit integer or a pointer.
+ * user. The type helps the receiver to multiplex different message types.
+ * The content can either be provided as a 32-bit integer or a pointer.
+ *
+ * Some message types are predefined; for example, @ref
+ * GNRC_NETAPI_MSG_TYPE_RCV & co are defined. These are fixed in the sense that
+ * registering for a particular set of messages (for the above, e.g. @ref
+ * gnrc_netreg_register) will use these message types. Threads that do nothing
+ * to receive such messages need not worry can safely use the same numbers for
```
Comment (kaspar030) [inline core/include/msg.h:25]:
```suggestion
 * to receive such messages need can safely use the same numbers for
```

**Comment by kaspar030** [comment]
LGTM, nice clarifications. Only the "may be considered an error", maybe needs some addition, or remove it? What do you think?

File: core/include/thread_flags.h (hunk 3 above)
Comment (chrysn) [reply core/include/thread_flags.h:105]: That was originally intentional to set it off from the others -- revisiting it, given the others don't have newlines between them, added.

**Comment by chrysn** [comment]
All applied, including an addition on best practice to log. (Although I think I'm more on the side of "log with extreme prejudice").

I'd advocate declaring this clearly as an error (not merely "usage error") not for the behavior when receiving an unexpected message, but for how it can go wrong worse: Whichever component sends a thread messages that the recipient does not expect might also hit a message type the recipient *does* expect, and then the recipient might free what's behind that pointer (if the semantics are similar to what is done in GNRC), and generally we're in deep UB and potential-security-issue land there. (Say the sender thinks of the message as a pointer to some user input buffer, but the recipient thinks of it as a struct and calls through a pointer in there...)

**Comment by chrysn** [comment]
Can I squash this?

**Comment by fjmolinas** [comment]
@kaspar030 is this Ok to squash?

**Comment by benpicco** [comment]
Please squash!

File: core/include/thread_flags.h [hunk 6]
```diff
@@ -98,6 +101,19 @@ extern "C" {
...
+ *
+ * This bit mask is set for all thread flag bits that are predefined in RIOT.
+ * Flags within this set may be set on a thread by the operating system without
+ * the thread soliciting them (though not all are; for example, @ref
+ * THREAD_FLAG_TIMEOUT is not).
+ *
+ * When using custom flags, asserting that they are not in this set can help
+ * avoid conflict with future additions to the predefined flags.
+ */
+#define THREAD_FLAG_PREDEFINED      (THREAD_FLAG_MSG_WAITING | THREAD_FLAG_TIMEOUT)
```
Comment (benpicco) [inline core/include/thread_flags.h:116]:
```suggestion
#define THREAD_FLAG_PREDEFINED_MASK (THREAD_FLAG_MSG_WAITING | THREAD_FLAG_TIMEOUT)
```

small nitpick: this should make it clearer that this is not an individual flag - feel free to squash directly

**Review by benpicco — APPROVED** [review]

**Review by kaspar030 — APPROVED** [review]
ACK.

**Comment by chrysn** [comment]
Thank you!

Now that these semantics are clarified I can go ahead and send zero-sized types (like exclusive access to some globals) between threads on flags :-)

**Pull request merged**